src/transformers/models/patchtst/modeling_patchtst.py

```diff
@@ -0,0 +1,2103 @@
+    else:
+        noise = torch.rand(
+            batch_size, num_channels, sequence_length, device=device
+        )  # noise in [0, 1], bs x num_channels x L
+
+    mask = torch.ones(
+        batch_size, num_channels, sequence_length, device=device
+    )  # mask: [bs x num_channels x num_patch]
+    mask[:, :, :len_keep] = 0
+
+    # sort noise for each sample
+    ids_shuffle = torch.argsort(noise, dim=-1)  # ascend: small is keep, large is remove
+    ids_restore = torch.argsort(ids_shuffle, dim=-1)  # ids_restore: [bs x num_channels x L]
+
+    mask = torch.gather(mask, dim=-1, index=ids_restore)
+    mask = mask.unsqueeze(-1).repeat(1, 1, 1, num_features)  # mask: [bs x num_channels x num_patches x patch_length]
+    if unmasked_channel_indices is not None:
+        mask[:, unmasked_channel_indices, :, :] = 0
+
+    inputs_mask = inputs.masked_fill(mask.bool(), mask_value)
+    return inputs_mask, mask[..., 0]
+
+
+def forecast_masking(
+    inputs: torch.Tensor,
+    patch_lengths: list,
+    mix_ratio: list = None,
+    unmasked_channel_indices: list = None,
+    mask_value: int = 0,
+    seed_number: Optional[int] = None,
+):
+    """Forecast masking that masks the last K patches where K is from the patch_lengths list.
+    For every batch, distribute the patch lengths based on mix_ratio and ignore masks for column indices mentioned in
+    unmasked_channel_indices.
+
+    Parameters:
+        inputs (`torch.Tensor`):
+            Input of shape `(bs, num_channels, num_patch, patch_len)` or `(bs, tsg1, tag2, num_channels, num_patch,
+            patch_len)`
+        patch_lengths (`list`):
+            List of patch lengths to mask in the end of the data.
+        mix_ratio (`list`, *optional*):
+            List of weights to use for each patch length. For Ex. if patch_lengths is [5,4] and mix_ratio is [1,1],
+            then equal weights to both patch lengths. Defaults to None.
+        unmasked_channel_indices (`list`, *optional*):
+            Control Variable channel indices. These channels will not be masked. Defaults to None.
+        mask_value (`int`, *optional* defaults to 0):
+            Value to use for masking. Defaults to 0.
+        seed_number (`int`, *optional*):
+            Value to set for the random seed.
+
+    Returns:
+        `tuple(torch.Tensor)`: inputs_mask, masked input, same shape as inputs Tensor and Mask tensor of shape `(bs,
+        num_channels , num_patch)` or `(bs, tsg1, tsg2, num_channels, num_patch)`
+    """
+    if seed_number:
+        set_seed(seed_number)
+
+    if mix_ratio is None:
+        mix_ratio = [1 for _ in patch_lengths]
+
+    batch_size, num_channels, sequence_length, num_features = inputs.shape
+    mask = torch.zeros(batch_size, num_channels, sequence_length, device=inputs.device)
+
+    t_list = []
+    total_length = 0
+    total_ratio = sum(mix_ratio)
+
+    for i, j in zip(patch_lengths, mix_ratio):
```

Comment:
would be great to use more declarative variable names here instead of `i` and `j`